### PR TITLE
Java-codegen: add support for Generic Maps.

### PR DESCRIPTION
**Pull request by remyhaemmerle-da**
This PR advances the state of Generic Maps #2256.

In this PR we use `LinkedHashMap` instead `HashMap` in the java bindings. 
Hence we can preserves order of map entries,  when (un)serializing value in protobuf.

### Pull Request Checklist

- [X] Read and understand the [contribution guidelines](https://github.com/digital-asset/daml/blob/master/CONTRIBUTING.md)
- [X] Include appropriate tests
- [X] Set a descriptive title and thorough description
- [X] Add a reference to the [issue this PR will solve](https://github.com/digital-asset/daml/issues), if appropriate
- [ ] Add a line to the [release notes](https://github.com/digital-asset/daml/blob/master/unreleased.rst), if appropriate
- [ ] Normal production system change, include purpose of change in description

NOTE: CI is not automatically run on non-members pull-requests for security
reasons. The reviewer will have to comment with `/AzurePipelines run` to
trigger the build.

### CODE REVIEW

File: language-support/java/bindings/src/main/java/com/daml/ledger/javaapi/data/DamlList.java [hunk 1]
```diff
@@ -33,6 +32,18 @@ public DamlList(@NonNull Value...values) {
...
+    public static <T> Collector<T, ArrayList<Value>, DamlList> collector(Function<T, Value> valueMapper) {
+        return Collector.of(
+                ArrayList::new,
+                (acc, entry) -> acc.add(valueMapper.apply(entry)),
+                (left, right) -> {
+                    left.addAll(right);
+                    return left;
+                },
+                DamlList::new
+        );
```
Comment (SamirTalwar) [inline language-support/java/bindings/src/main/java/com/daml/ledger/javaapi/data/DamlList.java]: Again, we can use:

```java
return collectingAndThen(Collectors.toList(), DamlList::new);
```

File: language-support/java/bindings/src/main/java/com/daml/ledger/javaapi/data/DamlMap.java [hunk 2]
```diff
@@ -7,23 +7,37 @@
...
+
+        return Collector.of(
+                HashMap::new,
+                (acc, entry) -> acc.put(keyMapper.apply(entry), valueMapper.apply(entry)),
+                (left, right) -> {
+                    left.putAll(right);
+                    return left;
+                },
+                DamlMap::new
+        );
```
Comment (SamirTalwar) [inline language-support/java/bindings/src/main/java/com/daml/ledger/javaapi/data/DamlMap.java]: You get the idea.

File: language-support/java/bindings/src/main/java/com/daml/ledger/javaapi/data/DamlGenMap.java [hunk 3]
```diff
@@ -7,12 +7,11 @@
...
-import java.util.Collections;
-import java.util.List;
-import java.util.Objects;
-import java.util.StringJoiner;
+import java.util.*;
+import java.util.function.Function;
+import java.util.stream.Collector;
 
-public class DamlGenMap extends Value {
+final public class DamlGenMap extends Value {
```
Comment (SamirTalwar) [inline language-support/java/bindings/src/main/java/com/daml/ledger/javaapi/data/DamlGenMap.java]:
```suggestion
public final class DamlGenMap extends Value {
```

File: language-support/java/bindings/src/main/java/com/daml/ledger/javaapi/data/DamlGenMap.java [hunk 4]
```diff
@@ -22,15 +21,30 @@ public DamlGenMap(java.util.Map<Value, Value> value) {
         this.value = value;
```
Comment (SamirTalwar) [inline language-support/java/bindings/src/main/java/com/daml/ledger/javaapi/data/DamlGenMap.java]: Is `DamlGenMap` supposed to be immutable? If it is, we might want to clone the map here (or make this private, and make a public factory method that clones it).

File: language-support/java/bindings/src/main/java/com/daml/ledger/javaapi/data/DamlGenMap.java [hunk 5]
```diff
@@ -22,15 +21,30 @@ public DamlGenMap(java.util.Map<Value, Value> value) {
...
+                (left, right) -> {
+                    left.putAll(right);
+                    return left;
+                },
+                DamlGenMap::new
+        );
+    }
+
     public @Nonnull
     java.util.Map<Value, Value> getMap() { return value; }
```
Comment (SamirTalwar) [inline language-support/java/bindings/src/main/java/com/daml/ledger/javaapi/data/DamlGenMap.java]: If the map is supposed to be immutable, might wanna return `Collections.unmodifiableMap(value)` here.

File: language-support/java/bindings/src/main/java/com/daml/ledger/javaapi/data/DamlGenMap.java (hunk 4 above)
Comment (remyhaemmerle-da) [reply language-support/java/bindings/src/main/java/com/daml/ledger/javaapi/data/DamlGenMap.java]: It is not suppose to be immutable. At least I copy from `DamlList` where the underlying list is mutable.
@gerolf-da should I make `DamlList`, `DamlMap`, and `DamlGenMap` immutable ?

File: language-support/java/bindings/src/main/java/com/daml/ledger/javaapi/data/DamlList.java (hunk 1 above)
Comment (remyhaemmerle-da) [reply language-support/java/bindings/src/main/java/com/daml/ledger/javaapi/data/DamlList.java]: This is not so simple. I have to map the list with `valueMapper`. Not sure it is worthwhile to make that following your approach. 

Do you prefer I drop the collector for List ? It is true that is not very usefull. I put it to here to follow the same pattern as Maps.

File: language-support/java/bindings/src/main/java/com/daml/ledger/javaapi/data/DamlGenMap.java [hunk 6]
```diff
@@ -22,15 +21,30 @@ public DamlGenMap(java.util.Map<Value, Value> value) {
...
+
+        return Collector.of(
+                LinkedHashMap::new,
+                (acc, entry) -> acc.put(keyMapper.apply(entry), valueMapper.apply(entry)),
+                (left, right) -> {
+                    left.putAll(right);
+                    return left;
+                },
+                DamlGenMap::new
+        );
```
[the remark this answers is not in the document]
Comment (remyhaemmerle-da) [reply language-support/java/bindings/src/main/java/com/daml/ledger/javaapi/data/DamlGenMap.java]: I do not find that much better. 
Your comment  "Not sure whether this is better or worse." does not really help me to choose.

File: language-support/java/bindings/src/main/java/com/daml/ledger/javaapi/data/DamlList.java [hunk 7]
```diff
@@ -6,11 +6,10 @@
 import com.digitalasset.ledger.api.v1.ValueOuterClass;
 import org.checkerframework.checker.nullness.qual.NonNull;
 
-import java.util.ArrayList;
-import java.util.Arrays;
-import java.util.List;
-import java.util.Objects;
+import java.lang.reflect.Array;
```
[the remark this answers is not in the document]
Comment (remyhaemmerle-da) [reply language-support/java/bindings/src/main/java/com/daml/ledger/javaapi/data/DamlList.java]: fixed.

File: language-support/java/bindings/src/main/java/com/daml/ledger/javaapi/data/DamlGenMap.java (hunk 5 above)
Comment (remyhaemmerle-da) [reply language-support/java/bindings/src/main/java/com/daml/ledger/javaapi/data/DamlGenMap.java]: In fact if we do not copy the map given as argument in the constructor, this is pretty useless.

File: language-support/java/bindings/src/main/java/com/daml/ledger/javaapi/data/DamlGenMap.java (hunk 4 above)
Comment (gerolf-da) [reply language-support/java/bindings/src/main/java/com/daml/ledger/javaapi/data/DamlGenMap.java]: Yes, I was looking at `DamlList` recently in another context and I noticed it there too. Ideally we copy the map (the values themselves should be already immutable).

File: language-support/java/bindings/src/main/java/com/daml/ledger/javaapi/data/DamlList.java (hunk 1 above)
Comment (SamirTalwar) [reply language-support/java/bindings/src/main/java/com/daml/ledger/javaapi/data/DamlList.java]: Makes sense. You can use `Collectors.mapping` but it’s up to you whether you find this cleaner or not.

File: language-support/java/bindings/src/main/java/com/daml/ledger/javaapi/data/DamlGenMap.java [hunk 8]
```diff
@@ -13,12 +13,24 @@
...
-    private final java.util.Map<Value, Value> value;
+    /**
+     * The map that is passed to this constructor must not be changed
+     * once passed.
+     */
+    private DamlGenMap(@NonNull Map<@NonNull Value, @NonNull Value> value) {
+        this.map = value;
+    }
+
+    private static @Nonnull DamlGenMap fromPrivateMap(@NonNull Map<@NonNull Value, @NonNull Value> map){
```
Comment (SamirTalwar) [inline language-support/java/bindings/src/main/java/com/daml/ledger/javaapi/data/DamlGenMap.java]: Did you mean to use both `Nonnull` and `NonNull`?

File: language-support/java/bindings/src/main/java/com/daml/ledger/javaapi/data/DamlGenMap.java (hunk 3 above)
Comment (remyhaemmerle-da) [reply language-support/java/bindings/src/main/java/com/daml/ledger/javaapi/data/DamlGenMap.java]: fixed

File: language-support/java/bindings/src/main/java/com/daml/ledger/javaapi/data/DamlGenMap.java (hunk 8 above)
Comment (remyhaemmerle-da) [reply language-support/java/bindings/src/main/java/com/daml/ledger/javaapi/data/DamlGenMap.java]: fixed

**Pull request merged**